### DISCUSSION TITLE
SubOrchestrator is not completing the execution.

### DISCUSSION THREAD

**Discussion by aminch18** [Azure/azure-functions-durable-extension#1970] [Q&A]
##### I purged the all instances history, but is like the SubOrchestrator began in a loop and never finish it.

![image](https://user-images.githubusercontent.com/38215834/136247880-561b4058-85e4-41d8-b9d8-670e8941afe9.png)

##### The final sequence diagram should be like this:
![image](https://user-images.githubusercontent.com/38215834/136248201-99bd4b1d-0821-457e-a5ac-ddaf5ef1f236.png)

##### Could you help me with this?

**Comment by cgillum** [comment]
For help with debugging these kinds of issues, it's best if you can share some additional information, such as where your code is running (local? Azure?), the language you're using, what is the instance ID of the impacted orchestration, and whether the [logs](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-diagnostics) contain any relevant information. Can you share more information about the sub-orchestration which isn't completing?

**Comment by aminch18** [comment]
#### Dear @cgillum,

#### I am using C# and .NETCore 3.1, Azure Function Version as v3, and the functions were deployed in Azure with Elastic Premium EP1 using 4 instances.

#### In logs, I saw that no exception have been thrown, and in fact the MigrateUserAlertActivity is storing correctly the data in Cosmos DB, but doesn't finish. And for this reason is not doing the next iteration this is the code of SubOrchestrator:

#### Orchestrator code:
![image](https://user-images.githubusercontent.com/38215834/136252552-c5b4cd32-2f95-4cc4-a100-4ba6aa07b8d7.png)

#### SubOrchestrator code:
![image](https://user-images.githubusercontent.com/38215834/136252588-431a3fce-be3d-4076-90ab-6f47b2f6e1f0.png)
![image](https://user-images.githubusercontent.com/38215834/136252611-dbc13161-1017-437b-9f98-7b275edb2e23.png)

#### Thanks in advance 😊

**Reply by cgillum** [reply]
This might be an internal execution problem. You mentioned this is running in Azure...would you be willing to share with me the orchestration instance ID (the value returned by `IDurableOrchestrationClient.StartNewAsync(...)`) of the parent orchestration (MigrationOrchestrator) that is stuck and a time range showing when you recently observed this issue?

**Reply by aminch18** [reply]
Dear @cgillum, I tried some different things:

1. Create a new Hub for the durable functions in order to use a new instance history, trying to use new hub in order to avoid corrupt data.
2. I added more logs in order to check the execution timeline and the Sub Orchestrator should be marked as **Completed** in the instances history, but the runtime status it is maintained as running.

For example:
History of instances:
![image](https://user-images.githubusercontent.com/38215834/136369040-f45c3e25-378f-44ad-9dee-13c4db2a2859.png)
Last execution
![image](https://user-images.githubusercontent.com/38215834/136368983-be52208c-a8ba-486a-9471-ee6940a714de.png)

As we can see in the latest image, the state is Completed but, it is not raising the Execution Completed event and the orchestrator have been stucked and for this reason it'll never do the next iteration of the loop.

Therefore, I am not able to know what is happening. 

Orchestrator instance ID for this last example: 17ae4314db8c45efb85fac1ce363c9ab
Sub orchestrator instance ID for this last example: 8d1d573d74d349c5af4109fdc9457ab7:0

Thanks in advance.

**Reply by cgillum** [reply]
Thanks for these details @aminch18. I was able to find some telemetry for your orchestrations and can confirm what you're seeing. I was able to see that your sub-orchestrator completed executing 4 activity functions, (MigrateAlertConfigurationsActivity, GetUserAlertsCount, MigrateUserAlertsActivity, and then another MigrateUserAlertsActivity). It looks like several minutes after this you terminated these orchestrations, presumably because they weren't completing properly?

Nothing obvious stands out to me about why this is happening, assuming the code you posted above is accurate. I did notice some warnings about sub-orchestration completion messages being lost due to the parent being terminated, but those events were for different times and instance IDs (6dc0a3c220b04300a95c34c54716fc48, and 6f1eb0177dcc4a02ad267e6599c3710c). 

However, I noticed that you're using an outdated version of the Durable Functions extension (v2.4.3). We regularly make important reliability and diagnostic fixes and its possible that either you're running into an issue that has been fixed in a recent update or that updating can give us more detailed telemetry on what's happening. Can you try updating to a newer version (the latest is v2.5.1) to see if the issue is still reproducible?

**Reply by aminch18** [reply]
Dear @cgillum, this is still happening even though I updated  Durable Functions extensions. I created a new host, but it is still not raising the events internally, and the sub orchestrator is not doing the loop properly.

**Reply by cgillum** [reply]
Thanks for checking, @aminch18. Do you have another orchestration instance ID that I can look at from after you upgraded the extension version?

**Reply by RDavis3000** [reply]
@aminch18 What happened next? I'm experiencing a similar problem!

**Reply by aminch18** [reply]
@RDavis3000, I've resolved this basically deleting my Azure Function and recreate it, I could not find the root cause, it seems that some data was corrupted inside the AZ Storage related to the AZ Function.

**Reply by aminch18** [reply]
@cgillum Thanks for all your help.
